### PR TITLE
Add GrantConditionOnAmmo

### PR DESCRIPTION
**Pull request by jrb0001**

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs]: This is probably copy pasta from `GrantConditionOnDamageState.cs` and not what this actually does.

**Review by Mailaender — CHANGES_REQUESTED** [review]

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs
Comment (jrb0001) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs]: fixed.

**Review by Mailaender — DISMISSED** [review]

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs]: Creating these objects every tick isn't great for memory churn.  Please create these once in the ctor, and then update the values each tick.

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs]: Can you please also add a comment `// TODO: Update these to account for disabled traits once AmmoPool is made conditional`

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs
Comment (jrb0001) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs]: done.

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs
Comment (jrb0001) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnAmmo.cs]: done.

**Comment by pchote** [comment]
Could you outline some of the ways that you plan to use the `Expression`?

**Comment by pchote** [comment]
The discussion in #13057 is particularly relevant here: under the proposed system `AmmoPool` would directly expose `Count` and `Capacity` variables that could be used directly in other conditions, instead of requiring a mediating trait.

**Comment by jrb0001** [comment]
The problem with integer conditions is that they rely on granting a condition multiple times iirc. This is a huge performance hit if ammopools have thousands of ammo.

I use the expression to grant a condition if the ammo pool is empty although I might also need it for other values.

**Comment by jrb0001** [comment]
rebased.

@Mailaender It looks like GitHub decided to dismiss your review automatically.

**Comment by Mailaender** [comment]
I haven't really tested it to be honest. It was just a quick skim.

**Comment by pchote** [comment]
> The problem with integer conditions is that they rely on granting a condition multiple times iirc. This is a huge performance hit if ammopools have thousands of ammo.

No, they expose their values directly.

**Comment by jrb0001** [comment]
hmm, then how do I grant such a condition? `ConditionManager.GrantCondition(Actor self, string condition, int duration = 0)` doesn't allow it.

**Comment by pchote** [comment]
The plumbing doesn't exist yet.  That is what #13057 is for.

**Comment by jrb0001** [comment]
Then it should be obvious that there is no other way to solve it at the moment. And the history of that PR tells me that it won't be merged in the near future. Do you want me to close my PR?

**Comment by pchote** [comment]
Like with #13847, I think it would be unwise to merge a temporary workaround when we have a general solution that would supersede it if not for lack of manpower.  It would be much better if you could help towards getting that implemented.

**Comment by jrb0001** [comment]
Ok, I will keep it downstream until the integer conditions are completely done and there is an official solution. I can implement it properly but there is no point in working on it until the plumbing PR is merged.

**Comment by reaperrr** [comment]
@jrb0001 We decided to add a local stacked conditions implementation to `AmmoPool` as temporary solution in #13538, I guess that might already be enough for some of the use-cases this PR wanted to adress.

**Comment by jrb0001** [comment]
The use-case was to grant conditions based on the ammo level. Even though only full and empty were needed, I attempted to implement a generic solution. #13538 should contain the required changes but it doesn't expose the ammo limit afaik.

**Comment by GraionDilach** [comment]
You're wrong - gladly, my request about exactly that was encouraged, see https://github.com/OpenRA/OpenRA/pull/13538#pullrequestreview-69615367.

**Pull request closed without merging**
